### ISSUE TITLE
Does NCCL Allreduce kernels slowdown the computation kernels of GPU?

### ISSUE THREAD

**Issue by ihchoi12** [NVIDIA/nccl#689]
Hello,

I'm trying to compare training speed between using 1 node and using 2 nodes (one GPU per node). 

From 1 node training, back-propagation (calculate gradients & update parameters) takes about 57ms with full GPU utilization.
<img width="915" alt="image" src="https://user-images.githubusercontent.com/37887404/171182342-202eeb70-45d9-4e11-86a9-141261b1f37c.png">

From 2 nodes training (with the same training setup), back-propagation takes about 237ms with full GPU utilization. 
<img width="968" alt="image" src="https://user-images.githubusercontent.com/37887404/171182700-40948d6e-7f8e-4f58-99f8-c9ab59bf0b42.png">

I believe the GPU is working on NCCL Allreduce operations in training with 2 nodes, and that's why the back-propagation takes much longer than the 1 node case. Does it mean that NCCL may slow down the overall training speed?

Additionally, I observe the much higher CPU utilization in the training with 2 nodes. Could you clarify why it is so? Is it because of the busy-waiting thread for the cudaEventSynchronize API call?

**Comment by sjeaugey** [comment]
Did you try to run NCCL after backprop and see what performance you get and how long NCCL takes to perform the allreduce? I can't see the NCCL time on the profiling.

Also in the 2 nodes screenshot, it seems to be the warmup phase instead of iteration 0 .. would it look the same on iteration 0?

NCCL is running on the GPU, so overlapping the NCCL communication and backprop can be tricky. The backprop kernels need to be tuned to not use the full GPU and leave space for NCCL, otherwise it could result is worse performance. I'm not expert in backprop though so I can only explain the theory, not the details.

Regarding CPU usage, indeed NCCL is using a busy-polling thread per NCCL rank (8 if running on 8 GPUs), waiting for the GPU to send data. We are not using `cudaEvent*`, just memory polling. Now, those threads should only be active when there is NCCL communication in progress which should not take long unless your network is very slow.

So I guess my first question would be how many allreduce operations do you perform during backprop, and on what size? How much time should that take given your hardware platform (running NCCL perf tests would tell you that)? Just to get a basic idea of what your compute time is, versus your communication time... there's not point optimizing the overlap if your NCCL time is 10x higher than your backprop; also I'd like to make sure NCCL is working at the expected speed and not taking 10x more than it should (hence blocking SMs longer than it should).

**Comment by sjeaugey** [comment]
> By default Horovod design which I'm testing, looks like the allreduce of calculated gradients is being performed in parallel with the backprop of the remaining layers, which is why we cannot see the NCCL time separately

I'm surprised. You should see NCCL operations in NVTX on a separate line. Now if you're using Horovod, I believe there is a way to tune horovod to perform a single big allreduce at the end. I'm not an expert in horovod tuning but setting `HOROVOD_FUSION_THRESHOLD` to a large value should probably do that.

> How does NCCL decide how many kernels to launch for itself? Does it try to fully utilize remaining (idle) GPU resources as much as possible?

NCCL uses the minimal number of SMs to achieve the performance of the platform. If you have a PCI system with a single NIC, which can achieve up to 12-24 GB/s it should just launch 2 SMs. If you are on DGX with 8 NICs and a top speed of 190 GB/s, then it would use 16 SMs.

> One thing I'm doubtful is that, I indeed see lots of cudaEventSynchronize calls during the backprop.

I can't explain that. NCCL does not call cudaEventSynchronize. So the increase CPU usage could indeed be a side effect of some other issue.

I'm still interested in knowing what speed your platform can achieve between 2 nodes, and the size of the total allreduce -- it's important to do that basic math to have in mind what the NCCL operation should look like in the profiling (how large the NCCL operations should be).

**Comment by ihchoi12** [comment]
> You should see NCCL operations in NVTX on a separate line.

I thought NVTX labels are the labels that I add (like "FP", "BP" in my profile screenshot)? About the NCCL operations, I see this NCCL row (separately from NVTX) like this, and quite many ncclAllReduce operations are performed during one backprop, which a quite short running time for each operation. Are you referring to this line?
<img width="1038" alt="image" src="https://user-images.githubusercontent.com/37887404/171371896-1ca62faa-d76a-4bb8-98e1-3963f76857c2.png">

> I believe there is a way to tune horovod to perform a single big allreduce at the end. I'm not an expert in horovod tuning but setting HOROVOD_FUSION_THRESHOLD to a large value should probably do that.

Exactly! I was trying this approach. The issue is that setting the HOROVOD_FUSION_THRESHOLD to a large value does not effectively aggregate all the allreduce operations to one. Looks like it's due to the different types of the tensors in the model (tensor fusion works for the same type of tensors). Let me investigate it more and will update here if I get some insight!
 

> I'm still interested in knowing what speed your platform can achieve between 2 nodes, and the size of the total allreduce -- it's important to do that basic math to have in mind what the NCCL operation should look like in the profiling (how large the NCCL operations should be).

The size of the total allreduce should be the size of the DNN model, right? I'm running this test with ResNet50 which has [98 MB](https://keras.io/api/applications/) of tensors. I will run the NCCL perf tests as you suggested, and update here. Thanks!!

**Comment by ihchoi12** [comment]
Hi @sjeaugey, may I confirm one thing?

In the previous comment, you said

> NCCL uses the minimal number of SMs to achieve the performance of the platform. If you have a PCI system with a single NIC, which can achieve up to 12-24 GB/s it should just launch 2 SMs. If you are on DGX with 8 NICs and a top speed of 190 GB/s, then it would use 16 SMs.

Here, GB/s means "gigabytes per second", not "gigabit per second"? I'm asking because we generally use "gigabit per second" for the NIC speed.

**Comment by sjeaugey** [comment]
It is Gigabytes per seconds as is usually measured by application-level benchmarks. NCCL benchmarks sometimes reflect the speed of the network but not always. Also the NCCL bandwidth started as PCI bandwidth, then NVLink, both of which are using GB/s.

**Comment by ihchoi12** [comment]
Got it. Actually, there is one thing suspicious to me. 

In my current testbed, I'm using two AWS p3.x2large nodes (one GPU per node), and the network bandwidth is 10Gbps (1.25 GB/s). According to your previous comment (`If you have a PCI system with a single NIC, which can achieve up to 12-24 GB/s it should just launch 2 SMs`), the NCCL allreduce should launch a minimal number of kernels, probably even not using 1 whole SM. However, I'm observing that the NCCL kernels are dominating most of the GPU resources while it's active as I can see in my Nsight Systems profile.  
![Uploading image.png…]()
May I know in which part of the code NCCL is calculating the number of kernels to launch? Maybe I would like to benchmark the duration of NCCL allreduce operation depending on the number of kernels it launches.

**Comment by sjeaugey** [comment]
You can force `NCCL_MAX_NCHANNELS=2` but on p3d given you only have one NIC I'm pretty sure you are really only using 2 SMs (2 is the minimum).

What you see on the Nsight Systems profile does not mean NCCL is using most of the GPU compute resources. In fact, if you run a CUDA kernel on just one SM and the kernel is active 100% of the time, you'll see the GPU is at 100% in nvidia-smi. Of course it's not using a log of power since only 1 SM is active but I'm pretty sure the %usage report does not reflect what % of SMs are used.

**Comment by ihchoi12** [comment]
That's interesting! But then, it's still unclear why the computation stream (Default stream 7 in the profile) becomes much longer with the parallel NCCL kernels. It took just 50ms when I run the training on a single node (i.e., no NCCL allreduce at all), meaning that the computation itself takes 50ms (as shown in the profile below). 
![image](https://user-images.githubusercontent.com/37887404/171808309-bfabf64c-7082-4812-8b58-06b4728796d9.png)
In 2-node case where the computation is happening in parallel with the NCCL allreduce operations, it takes more than 200ms, and to me, it seems because the computation stream and NCCL stream are competing each other for the limited GPU resource.

Also, if the Nsight Systems report the GPU %usage of a single SM running, which SM's %usage does it report when there are multiple SMs running at the same time? For example, during backprop in my 2-node case, there are two SMs (Stream 13 for NCCL and Stream 7 for computation). Isn't it a bit weird that it reports the %usage of only one of SMs?

(Btw, thank you so much for answering and discussing this. It's really helpful!)

**Comment by sjeaugey** [comment]
> But then, it's still unclear why the computation stream (Default stream 7 in the profile) becomes much longer with the parallel NCCL kernels.

As I mentioned, I can't tell whether it's normal or not. CUDA scheduling can be tricky to get right and playing with Horovod tuning could help.

Now you mentioned you were running on p3d (not p3dn) and your NIC was 10Gb/s. This is really, really low. Performing a single allreduce at the end would take 80ms at best. Since it seems Horovod is splitting the allreduce into many small operations, it's probably taking twice that time. So even if NCCL was perfectly overlapped, you'd still see a bwd time of ~150ms which would defeat the purpose of running on 2 GPUs.

Given your backprop time (50ms) and gradient size (98MB), you need the NCCL allreduce to take at most 50ms, meaning 2*98/BW < 50ms. So BW > 4 GB/s. And that's with perfect overlap, which is usually not what we see. That also means you need ~20 GB/s to get 80% efficiency without overlap.

That's why I'm saying it's important to run NCCL perf tests and do some back-of-the-envelope math to know what BW you need before optimizing anything. You need to run on a single node with GPU connected through PCI or NVLink, and if you want to go beyond one node, you need a 100G NIC at the very minimum; 10G is not going to be enough.

**Comment by ihchoi12** [comment]
Hi @sjeaugey!

I tried the nccl-perf test using the same two p3 nodes, and trying to understand the result right now. 
When I generate the Nsight Systems profile from the perf test, I observe some interesting differences in the NCCL streams on GPU.

Previously when I see the profile of pytorch-horovod training script, the NCCL stream (Stream 13) was running some kernels called `ncclKernel_AllReduce_RING_LL_Sum_float(ncclWorkElem)` (shown in the above profile screenshot).
 
However, in the nccl-perf test, the NCCL stream (Stream 13) is performing `Memcpy DtoD` as shown here: 
<img width="1020" alt="image" src="https://user-images.githubusercontent.com/37887404/171876430-38bee788-8b0f-406d-a1fa-d2942df036e0.png">

I wonder why the same NCCL allreduce operation is showing this different behavior on this identical testbed (two p3 nodes). Also, in the nccl-perf case, how the GPU can directly perform the `Memcpy DtoD` operation while the other GPU is on the other node? Could you please help me to understand this?

**Comment by sjeaugey** [comment]
This is surprising. Looks like a 1-GPU allreduce (which is just a copy from the input buffer to the output buffer).

**Comment by ihchoi12** [comment]
Ah, the previous post was actually 1-GPU allreduce by my mistake (there were some problems in the mpi setup). Now, I got a reasonable result of 2-GPU allreduce and analyzing it. The previous post has been deleted after I solved the issue. Thanks, @sjeaugey!

**Comment by korsaka** [comment]
>> NCCL uses the minimal number of SMs to achieve the performance of the platform. If you have a PCI system with a single NIC, which can achieve up to 12-24 GB/s it should just launch 2 SMs. If you are on DGX with 8 NICs and a top speed of 190 GB/s, then it would use 16 SMs.
>  Excuse me, I have a question about how many SMs NCCL will use.  NCCL will launch one cuda block(640 threads at most) per channel.  There are 2048 threads per SM for A100，so in my option one SM can serve three NCCL channels not one, am i right?  Looking forward to you reply, thank you  @sjeaugey

**Comment by sjeaugey** [comment]
No, the reason we use 640 threads per SM is because we wouldn't get more performance with more threads. We're limited by the bandwidth one SM can achieve, not by the number of threads.

**Comment by korsaka** [comment]
> No, the reason we use 640 threads per SM is because we wouldn't get more performance with more threads. We're limited by the bandwidth one SM can achieve, not by the number of threads.


Got it，So one nccl block will occupy one SM exclusively?

**Comment by korsaka** [comment]
> yes, NCCL CTAs won't share an SM with compute kernels.


Got it， thank you very much

**Comment by autumn0207** [comment]
any update?

**Comment by jammycc** [comment]
> > So one nccl block will occupy one SM exclusively
> 
> yes, NCCL CTAs won't share an SM with compute kernels.

hi sjeaugey! Why do NCCL kernels not share SMs with compute kernels, even if there are leftover resources on the SM? Are there any configuration parameters for this?

**Comment by sjeaugey** [comment]
> Why do NCCL kernels not share SMs with compute kernels, even if there are leftover resources on the SM? 

NCCL kernels won't share kernels with others because NCCL needs to use all registers to reach peak bandwidth. So, there is no leftover resource on the SM.

**Comment by ColdPorridge** [comment]
> > Why do NCCL kernels not share SMs with compute kernels, even if there are leftover resources on the SM?
> 
> NCCL kernels won't share kernels with others because NCCL needs to use all registers to reach peak bandwidth. So, there is no leftover resource on the SM.

@sjeaugey hi, sjeaugey! I would like to ask a few more questions about this:

- How does NCCL ensure that the NCCL kernel has exclusive access to the SM? How does the thread block scheduler prevent other kernels from being allocated to the SM where the NCCL kernel is running? Is it achieved by let NCCL kernel using all registers on SM, making the thread block scheduler believe that the remaining register resources in that SM are insufficient for any other kernels, thus avoiding their allocation to this SM?

- If the number of completely idle SMs is less than the number of channels for the NCCL kernel, how does the thread block scheduler allocate SMs for the NCCL kernel? Does it make the NCCL kernel wait until there are enough completely idle SMs available?

**Comment by Graham1025** [comment]
> > > Why do NCCL kernels not share SMs with compute kernels, even if there are leftover resources on the SM?
> > 
> > 
> > NCCL kernels won't share kernels with others because NCCL needs to use all registers to reach peak bandwidth. So, there is no leftover resource on the SM.
> 
> [@sjeaugey](https://github.com/sjeaugey) hi, sjeaugey! I would like to ask a few more questions about how NCCL ensures that the NCCL kernel has exclusive access to the SM.
> 
> * How does the thread block scheduler prevent other kernels from being allocated to the SM where the NCCL kernel is running? Is this achieved by having the NCCL kernel use all the registers on the SM, making the thread block scheduler believe that the remaining register resources in that SM are insufficient for any other kernels, thus avoiding their allocation to this SM?
> * If the number of completely idle SMs is less than the number of channels for the NCCL kernel to be launched, how does the thread block scheduler allocate SMs for the NCCL kernel? Does it make the NCCL kernel wait until there are enough completely idle SMs available?

I have the same question, too.  looking forward to your reply！[@sjeaugey](https://github.com/sjeaugey)

**Comment by sjeaugey** [comment]
Indeed NCCL doesn't really prevent the scheduler from running other kernels concurrently, it's just a side effect from the fact that copy/reduce code in NCCL uses all registers for unrolling, causing the scheduler to effectively never schedule another kernel on the same SM.

As for scheduling all CTAs in a NCCL kernel at once, I would think that it should be possible for CUDA to launch only a fraction of CTAs (waiting for more CTAs to free up) and not waiting for enough CTAs to be free. I don't think we explicitly define our kernels as "cooperative", meaning that they need to launch at once.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
